src/ansi-c/goto-conversion/goto_clean_expr.cpp

```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
   }
   else if(expr.id() == ID_forall || expr.id() == ID_exists)
   {
+    quantifier_exprt &qex = to_quantifier_expr(expr);
+    code_expressiont code{qex.where()};
     DATA_INVARIANT(
-      !has_subexpr(expr, ID_side_effect),
-      "the front-end should check quantified expressions for side-effects");
+      !has_subexpr(expr, ID_side_effect) ||
+        (code.operands()[0].id() == ID_side_effect &&
+         code.operands()[0].get_named_sub()[ID_statement].id() ==
+           ID_statement_expression),
+      "quantifier must not contain side effects");
+
+    // Handle the case that quantifier body is a statement expression.
+    if(
+      code.operands()[0].id() == ID_side_effect &&
+      code.operands()[0].get_named_sub()[ID_statement].id() ==
+        ID_statement_expression)
+    {
+      goto_programt where;
+      goto_convert(code, where, mode);
+      where.compute_location_numbers();
+
+      natural_loops_mutablet natural_loops(where);
+      INVARIANT(
+        natural_loops.loop_map.size() == 0,
+        "quantifier must not contain side effects");
+
+      // `last` is the instruction corresponding to the last expression in the
+      // statement expression.
+      goto_programt::const_targett last;
```

Comment:
I'd suggest that you initialise this to `where.instructions.end()` and, after the loop, have a `DATA_INVARIANT` that `last != where.instructions.end()`. As is, `last->get_other()` could potentially be using an uninitialised iterator.